### PR TITLE
Add TypeScript REST API validation

### PR DESCRIPTION
**Pull request by tamim-khan**
## Notion ticket link
[TypeScript REST API input validation](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d?p=44f96cc7a8c045edb647bb266aa120eb)

## Implementation description
* Adding middleware to validate Request DTOs

## Steps to test
1. Requests with invalid DTO shape (invalid types or missing fields) should return `400 Bad Request`

## What should reviewers focus on?
* Is a 400 Response good enough or should we return the list of incorrect fields in the response? If so, in what format?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/middlewares/validators/authValidators.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Request, Response, NextFunction } from "express";
+import isValid from "./isValid";
+
+/* eslint-disable-next-line import/prefer-default-export */
```
Comment (tamim-khan) [inline backend/typescript/middlewares/validators/authValidators.ts:4]: Avoiding default-export so it's easier to add new validators to this file.

File: backend/typescript/middlewares/validators/entityValidators.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { Request, Response, NextFunction } from "express";
+import isValid from "./isValid";
+
+/* eslint-disable-next-line import/prefer-default-export */
```
Comment (tamim-khan) [inline backend/typescript/middlewares/validators/entityValidators.ts:4]: Same here

File: backend/typescript/middlewares/validators/isValid.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  if (type === "string") {
+    return typeof value === "string";
+  }
+  if (type === "boolean") {
+    return typeof value === "boolean";
+  }
+  if (type === "integer") {
+    return typeof value === "number" && Number.isInteger(value);
+  }
+  if (type === "string-array") {
```
Comment (sherryhli) [inline backend/typescript/middlewares/validators/isValid.ts:15]: This is just a nice-to-have but we could create separate `validatePrimitive` and `validateArray` functions, where `validateArray` is passed the type of its elements. This way we can easily validate arrays of any type. Up to you if you want to make this change now, if not, I can create a task for the future

File: backend/typescript/middlewares/validators/userValidators.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  }
+  if (!isValid(req.body.lastName, "string")) {
+    return res.sendStatus(400);
+  }
+  if (!isValid(req.body.email, "string")) {
+    return res.sendStatus(400);
+  }
+  if (!isValid(req.body.role, "string")) {
+    return res.sendStatus(400);
+  }
```
Comment (sherryhli) [inline backend/typescript/middlewares/validators/userValidators.ts:20]: Could reuse `updateUserDtoValidator` here

File: backend/typescript/middlewares/validators/authValidators.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+export const loginRequestValidator = async (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  if (!isValid(req.body.email, "string")) {
+    return res.sendStatus(400);
```
Comment (sherryhli) [inline backend/typescript/middlewares/validators/authValidators.ts:11]: Can we return an error message please? Same thing for the other validators

File: backend/typescript/middlewares/validators/isValid.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+type Type = "string" | "integer" | "boolean" | "string-array";
+
+const isValid = (value: any, type: Type): boolean => {
+  if (value === undefined || value === null) return false;
+
+  if (type === "string") {
```
Comment (sherryhli) [inline backend/typescript/middlewares/validators/isValid.ts:6]: Nit: I think switch case is a bit cleaner

**Review by sherryhli — CHANGES_REQUESTED** [review]
A couple of minor things and then you're good to go! Only required change is to add error message, everything else is optional

File: backend/typescript/middlewares/validators/userValidators.ts (hunk 4 above)
Comment (tamim-khan) [reply backend/typescript/middlewares/validators/userValidators.ts]: That could prematurely call `next()`

**Review by sherryhli — APPROVED** [review]
LGTM, 🚢 it!

**Pull request merged**